### PR TITLE
retry on {Node,Store}ID allocation

### PR DESCRIPTION
**Pull request by tbg**
The calls to generate Node and Store IDs should be retried when they fail.

Otherwise, Fatals like this can happen:

> node connected via gossip and verified as part of cluster "..."
> unable to allocate node ID: the descriptor for the first range is not available via gossip

### CODE REVIEW

**Comment by tamird** [comment]
Can you add a test for this?

**Comment by tamird** [comment]
LGTM but needs a test

File: server/node.go [hunk 1]
```diff
@@ -80,24 +89,50 @@ type Node struct {
...
+				status = retry.Continue
+			}
+			return status, util.Errorf("unable to allocate node ID: %s", err)
+		}
+		id = proto.NodeID(r.ValueInt())
+		return retry.Break, nil
+	})
 	if err != nil {
-		return 0, util.Errorf("unable to allocate node ID: %s", err)
+		log.Fatal(err)
```
Comment (bdarnell) [inline server/node.go]: Why did you move this? We already had errors being passed up the stack, so it seems better to continue to do that than to add a log.Fatal at a lower level.

**Comment by bdarnell** [comment]
LGTM

File: server/node.go (hunk 1 above)
Comment (tschottdorf) [inline server/node.go]: You're right. Moved it back up.

**Comment by tamird** [comment]
LGTM thanks for adding the test

**Comment by tschottdorf** [comment]
you're welcome.

**Comment by spencerkimball** [comment]
I think this is the wrong place to do the retry. The dist sender should not be returning an error if the first range isn't available on gossip. It should probably be retrying itself, not returning an error. That would necessitate that every client must do its own retries in case it runs into this same problem when connecting to a cluster where gossip doesn't have the first range descriptor.

**Comment by tamird** [comment]
@spencerkimball doesn't that presume that `allocateNodeID` will need to move down the stack? at this level we don't know the cause of the error.

**Comment by tschottdorf** [comment]
It's operating on a bare DistSender, and having something else down there that can cause infinite retries sounds like a pain. I'm happiest about retries at the top level, clients need retry logic anyways and they don't need to care about the specific reason, that's why we have `proto.Error`.

**Comment by spencerkimball** [comment]
@tamird I don't understand your comment. What do you mean by "down the stack"? My point is that this level shouldn't be dealing with retryable errors. There are plenty of retryable error types, currently all handled in dist_sender. Why not continue handling those there?

**Comment by spencerkimball** [comment]
@tschottdorf What do you mean a bare DistSender. This code is using a "db" client object. We also use the same "db" client object when doing a split or a merge. But I notice you haven't also made changes there to retry on retryable errors. Why not? The exact same error regarding lack of first range metadata in gossip might occur anywhere in the system which we use a "db" client. It's not reasonable to build retry loops for this kind of thing into all of them. If the dist sender can retry a retryable error, it should be retrying. Otherwise, why not just let all of the retryable errors propagate from the dist sender and wrap all of our client usage in retry loops?

**Comment by tamird** [comment]
@spencerkimball ah, so you're suggesting that this retry goes here: https://github.com/cockroachdb/cockroach/blob/master/kv/dist_sender.go#L663 ?

**Comment by tschottdorf** [comment]
@tamird @spencerkimball I think this slipped through the cracks here then: https://github.com/cockroachdb/cockroach/pull/1248/files#r31667485

**Comment by tamird** [comment]
Yup, seems so. @tschottdorf want to try adding that retry there and seeing what the fallout is?

**Comment by spencerkimball** [comment]
Yeah, we just weren't considering the possibility of lacking the first range descriptor. Let's just make sure that the resulting error is retryable and allow the result to trigger this retry loop.

**Comment by tschottdorf** [comment]
I'll back out this change and do the proper retry instead. Not sure I'll get to it today though (it's already late here), so if either of you is burning to get this done, feel free to just go for it.

**Comment by spencerkimball** [comment]
No rush!
